src/transformers/models/mega/modeling_mega.py

```diff
@@ -0,0 +1,2280 @@
+"""PyTorch MEGA model."""
+
+import math
+from typing import List, Optional, Tuple, Union
+
+import torch
+import torch.nn.functional as F
+import torch.utils.checkpoint
+from torch import nn
+from torch.nn import BCEWithLogitsLoss, CrossEntropyLoss, MSELoss
+
+from ...activations import ACT2FN
+from ...modeling_outputs import (
+    BaseModelOutputWithPoolingAndCrossAttentions,
+    CausalLMOutputWithCrossAttentions,
+    MaskedLMOutput,
+    MultipleChoiceModelOutput,
+    QuestionAnsweringModelOutput,
+    SequenceClassifierOutput,
+    TokenClassifierOutput,
+)
+from ...modeling_utils import PreTrainedModel
+from ...pytorch_utils import ALL_LAYERNORM_LAYERS
+from ...utils import (
+    add_code_sample_docstrings,
+    add_start_docstrings,
+    add_start_docstrings_to_model_forward,
+    logging,
+    replace_return_docstrings,
+)
+from .configuration_mega import MegaConfig
+
+
+logger = logging.get_logger(__name__)
+
+_CHECKPOINT_FOR_DOC = "mnaylor/mega-base-wikitext"
+_CONFIG_FOR_DOC = "MegaConfig"
+
+MEGA_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "mnaylor/mega-base-wikitext",
+    # See all Mega models at https://huggingface.co/models?filter=mega
+]
+
+
+# utility for causal LM masking in the format that Mega expects
+def generate_causal_mask(seq_len):
+    """
+    Tiny utility to generate a `seq_len` by `seq_len` causal mask, where 1 corresponds to *not masked* and 0
+    corresponds to *masked*
+
+    causal_mask[i][j] corresponds to whether token `i` can attend to token `j`
+    """
+    seq_ids = torch.arange(seq_len)
+    causal_mask = seq_ids[None, :].repeat(seq_len, 1) <= seq_ids[:, None]
+    return causal_mask.to(torch.long)
+
+
+# Positional embeddings
+# copied from original Mega code and renamed variables for better readability
+class SimpleRelativePositionalBias(nn.Module):
+    def __init__(self, config: MegaConfig):
+        super().__init__()
+        self.config = config
+        self.max_positions = self.config.max_positions if self.config.chunk_size < 0 else self.config.chunk_size
+        self.rel_pos_bias = nn.Parameter(torch.Tensor(2 * config.max_positions - 1))
+
+    def forward(self, seq_len):
+        if seq_len > self.max_positions:
+            raise ValueError("Sequence length {} going beyond max length {}".format(seq_len, self.max_positions))
+
+        # seq_len * 2 - 1
+        bias = self.rel_pos_bias[(self.max_positions - seq_len) : (self.max_positions + seq_len - 1)]
+        # seq_len * 3 - 1
+        tile = F.pad(bias, (0, seq_len))
+        # (seq_len * 3 - 1) * seq_len
+        tile = torch.tile(tile, (seq_len,))
+        tile = tile[:-seq_len]
+        # seq_len x (3 * seq_len - 2)
+        tile = tile.view(seq_len, 3 * seq_len - 2)
+        start = (2 * seq_len - 1) // 2
+        end = tile.size(1) - start
+        tile = tile[:, start:end]
+        return tile
+
+
+# rotary positional bias similar to RoPE
+# taken from original Mega repo
+class RotaryRelativePositionalBias(nn.Module):
```

Comment:
Also we recently had a problem with `padding = "left"` when using the rotary embedding ( see #22161 )
If this is similar to `RoPE`, we have it in transformers so using `# Copied from` would also work well and prevent from adding new code, but when I checked, it seems that we just compute a bias and add it to the query and key states, vs scale the query and key

Reply:
That's super interesting! It looks like the Mega implementation is susceptible to the same problem:

```python
from transformers import AutoTokenizer, MegaConfig, MegaModel
import torch
tok_left = AutoTokenizer.from_pretrained("mnaylor/mega-base-wikitext", padding_side="left")
inputs_left = tok_left(["This is an example tokenized input with padding"], return_tensors="pt", padding="max_length", max_length=25)

tok_right = AutoTokenizer.from_pretrained("mnaylor/mega-base-wikitext")
inputs_right = tok_right(["This is an example tokenized input with padding"], return_tensors="pt", padding="max_length", max_length=25)

model = MegaModel(MegaConfig(vocab_size=tok_right.vocab_size)).eval()
outputs_left = model(inputs_left['input_ids'], inputs_left['attention_mask'])
outputs_right = model(inputs_right['input_ids'], inputs_right['attention_mask'])

torch.allclose(outputs_left[0], outputs_right[0], atol=1e-5) # False
(outputs_left[0] - outputs_right[1]).max() # 0.3989
```

I'll dig into their fix and see if it will be easy to port over.